src/mscorlib/src/System/IO/BufferedStream.cs

```diff
@@ -641,7 +641,9 @@ public BufferedStream(Stream stream)
         if (t != null && t.Result == val)
             return t;
         
-        t = Task.FromResult<Int32>(val);
+        t = val == 0 ?
+            TaskCache.Zero : // Return a cached Task if EOF has been reached
+            Task.FromResult<int>(val);
```

Comment:
This, and the other places in this PR like it, is adding an extra branch to every read, and the only time it'll help avoid an allocation is on the very last read to the stream.  Typically streams are read from many times, especially in the case of a buffered stream, so this is adding such branches across all read calls in exchange for saving a single allocation for the lifetime of the stream.